In [1]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_0cc49a4247064ece92a2b2ec1f62d4e9.csv"
df=pd.read_csv(url,header=None)
df.head()

0                       1  \
0  EI        ATRINS-DONOR-521   
1  EI        ATRINS-DONOR-905   
2  EI        BABAPOE-DONOR-30   
3  EI       BABAPOE-DONOR-867   
4  EI      BABAPOE-DONOR-2817   

                                                   2  
0  CCAGCTGCATCACAGGAGGCCAGCGAGCAGGTCTGTTCCAAGGGCC...  
1  AGACCCGCCGGGAGGCGGAGGACCTGCAGGGTGAGCCCCACCGCCC...  
2  GAGGTGAAGGACGTCCTTCCCCAGGAGCCGGTGAGAAGCGCAGTCG...  
3  GGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGCGGGGCTT...  
4  GCTCAGCCCCCAGGTCACCCAGGAACTGACGTGAGTGTCCCCATCC...

In [2]:
df.shape

(3190, 3)

In [3]:
df=df.rename(columns={0:"Class"})
y=df[["Class"]]
y.groupby("Class")["Class"].count()

Class
EI     767
IE     768
N     1655
Name: Class, dtype: int64

In [4]:
classes = df.iloc[:,0]
print(classes[:5])

0    EI
1    EI
2    EI
3    EI
4    EI
Name: Class, dtype: object


In [5]:
sequences = list(df.iloc[:, 2])
dataset = {}

for i, seq in enumerate(sequences):
    nucleotides = list(seq)
    nucleotides = [x for x in nucleotides if x != '\t']
    nucleotides.append(classes[i])
    dataset[i] = nucleotides
    
print(dataset[0])

['C', 'C', 'A', 'G', 'C', 'T', 'G', 'C', 'A', 'T', 'C', 'A', 'C', 'A', 'G', 'G', 'A', 'G', 'G', 'C', 'C', 'A', 'G', 'C', 'G', 'A', 'G', 'C', 'A', 'G', 'G', 'T', 'C', 'T', 'G', 'T', 'T', 'C', 'C', 'A', 'A', 'G', 'G', 'G', 'C', 'C', 'T', 'T', 'C', 'G', 'A', 'G', 'C', 'C', 'A', 'G', 'T', 'C', 'T', 'G', 'EI']


In [6]:
X = pd.DataFrame(dataset)
X

0    1    2    3    4    5    6    7    8    9     ... 3180 3181 3182 3183  \
0     C    A    G    G    G    C    C    C    T    A  ...    C    C    T    T   
1     C    G    A    G    C    A    C    C    G    A  ...    C    T    G    A   
2     A    A    G    G    T    G    T    C    G    G  ...    T    G    T    A   
3     G    C    G    C    C    A    T    T    C    C  ...    G    G    T    A   
4     C    C    T    T    A    C    T    C    G    T  ...    C    A    T    A   
..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
56    T    C    G    C    C    A    G    G    G    G  ...    C    G    C    T   
57    C    C    A    C    C    C    C    G    T    A  ...    G    T    T    C   
58    T    G    T    C    C    A    C    G    C    G  ...    A    A    G    T   
59    G    C    G    C    T    G    A    G    C    A  ...    C    T    T    T   
60   EI   EI   EI   EI   EI   EI   EI   EI   EI   EI  ...    N    N    N    N   

   3184 3185 3186 3187 3188 3189  
0     A    T    G    T    A    A  
1     T    C    A    C    T    G  
2     C    T    G    T    T    G  
3     A    C    C    C    C    C  
4     A    T    T    G    T    T  
..  ...  ...  ...  ...  ...  ...  
56    A    T    G    T    A    C  
57    C    C    C    G    C    A  
58    T    T    T    T    A    T  
59    A    T    G    C    A    T  
60    N    N    N    N    N    N  

[61 rows x 3190 columns]

In [7]:
X=X.transpose()
X.head()

0  1  2  3  4  5  6  7  8  9   ... 51 52 53 54 55 56 57 58 59  60
0  C  C  A  G  C  T  G  C  A  T  ...  G  C  C  A  G  T  C  T  G  EI
1  A  G  A  C  C  C  G  C  C  G  ...  T  G  C  C  C  C  C  G  C  EI
2  G  A  G  G  T  G  A  A  G  G  ...  A  C  G  G  G  G  A  T  G  EI
3  G  G  G  C  T  G  C  G  T  T  ...  G  T  T  T  T  C  C  C  C  EI
4  G  C  T  C  A  G  C  C  C  C  ...  C  T  T  G  A  C  C  C  T  EI

[5 rows x 61 columns]

In [8]:
X=X.iloc[:,0:60]
X.head()

0  1  2  3  4  5  6  7  8  9   ... 50 51 52 53 54 55 56 57 58 59
0  C  C  A  G  C  T  G  C  A  T  ...  A  G  C  C  A  G  T  C  T  G
1  A  G  A  C  C  C  G  C  C  G  ...  G  T  G  C  C  C  C  C  G  C
2  G  A  G  G  T  G  A  A  G  G  ...  C  A  C  G  G  G  G  A  T  G
3  G  G  G  C  T  G  C  G  T  T  ...  G  G  T  T  T  T  C  C  C  C
4  G  C  T  C  A  G  C  C  C  C  ...  C  C  T  T  G  A  C  C  C  T

[5 rows x 60 columns]

In [9]:
from sklearn.preprocessing import OneHotEncoder
enc=OneHotEncoder(handle_unknown="ignore")
X=pd.DataFrame(enc.fit_transform(X).toarray())
X.head()

0    1    2    3    4    5    6    7    8    9    ...  277  278  279  280  \
0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
1  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0  0.0   
2  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  1.0  0.0  0.0   
4  0.0  0.0  0.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  1.0  0.0  0.0   

   281  282  283  284  285  286  
0  1.0  0.0  0.0  1.0  0.0  0.0  
1  0.0  0.0  1.0  0.0  0.0  0.0  
2  1.0  0.0  0.0  1.0  0.0  0.0  
3  0.0  0.0  1.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  1.0  

[5 rows x 287 columns]

In [10]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)
print(X_train.shape)
print(X_test.shape)

(2392, 287)
(798, 287)


In [11]:
from sklearn import svm
import numpy as np
model1=svm.SVC(kernel="poly",C=2,
               random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
model1

SVC(C=2, kernel='poly', random_state=0)

In [12]:
from sklearn.metrics import accuracy_score
train_pred=model1.predict(X_train)
test_pred=model1.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9995819397993311
Accuracy on test data:  0.9711779448621554


In [13]:
from sklearn.metrics import confusion_matrix
print("Confusion matrix for test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix for test data: 
 [[209   6   1]
 [  3 171   5]
 [  2   6 395]]


In [14]:
from sklearn.multiclass import OneVsRestClassifier
model2 = OneVsRestClassifier(svm.SVC(kernel="poly",C=2,
                             random_state=0)).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
model2

OneVsRestClassifier(estimator=SVC(C=2, kernel='poly', random_state=0))

In [15]:
train_pred=model2.predict(X_train)
test_pred=model2.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9995819397993311
Accuracy on test data:  0.9724310776942355


In [16]:
print("Confusion matrix for test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix for test data: 
 [[209   6   1]
 [  2 174   3]
 [  3   7 393]]


In [17]:
from sklearn.multiclass import OneVsOneClassifier
model3 = OneVsOneClassifier(svm.SVC(kernel="poly",C=2,
                             random_state=0)).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
model3

OneVsOneClassifier(estimator=SVC(C=2, kernel='poly', random_state=0))

In [18]:
train_pred=model3.predict(X_train)
test_pred=model3.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9995819397993311
Accuracy on test data:  0.9711779448621554


In [19]:
print("Confusion matrix for test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix for test data: 
 [[209   6   1]
 [  3 171   5]
 [  2   6 395]]


#### Summary:
- SVM with "poly" kernel: 97.11% accuracy on test data.
- SVM with "poly" kernel wrapped inside OneVsRestClassifier: 97.24% accuracy on test data.
- SVM with "poly" kernel wrapped inside OneVsOneClassifier: 97.11% accuracy on test data.

### The following is an experiment unrelated to the models above.

UCI reported on its website that "This dataset has been developed to help evaluate a "hybrid" learning algorithm (KBANN) that uses examples to inductively refine preexisting knowledge. Using a "ten-fold cross-validation" methodology on 1000 examples randomly selected from the complete set of 3190, the following error rates were produced by various ML algorithms (all experiments run at the Univ of Wisconsin, sometimes with local implementations of published algorithms)."

|System|Neither|EI|IE|
|------|-------|--|--|
|KBANN|4.62%|7.56%|8.47%|
|BACKPROP|5.29%|5.74%|10.75%|
|PEBLS|6.86%|8.18%|7.55%|
|PERCEPTRON|3.99%|16.32%|17.41%|
|ID3|8.84%|10.58%|13.99%|
|COBWEB|11.80%|15.04%|9.46%|
|Nearest Neighbors|31.11%|11.65%|9.09%|

#### Let's randomly select 1000 samples from dataset and run a 10-fold cross validations on it as well. 

In [44]:
df=pd.concat([X,y],axis=1)
df1=df.sample(frac=1000/3190,random_state=0)
df1.shape

(1000, 288)

In [45]:
X1=df1.drop(["Class"],axis=1)
X1.head()

0    1    2    3    4    5    6    7    8    9    ...  277  278  279  \
982  0.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  0.0  ...  0.0  1.0  0.0   
702  0.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  1.0  0.0   
461  0.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
480  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  1.0  0.0  0.0   
298  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0  0.0  1.0   

     280  281  282  283  284  285  286  
982  0.0  0.0  1.0  0.0  0.0  0.0  0.0  
702  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
461  0.0  1.0  0.0  0.0  0.0  0.0  1.0  
480  0.0  0.0  0.0  0.0  0.0  0.0  1.0  
298  0.0  0.0  1.0  0.0  0.0  0.0  0.0  

[5 rows x 287 columns]

In [46]:
y1=df1[["Class"]]
y1.head()

Class
982    IE
702    EI
461    EI
480    EI
298    EI

In [47]:
from sklearn import model_selection
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import cross_val_predict
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [55]:
from sklearn import svm
model4=svm.SVC(kernel="poly",random_state=0)
scores=cross_val_score(model4,X1,np.asarray(y1).reshape(y1.shape[0],),cv=10)
print("Accuracy is: ",scores.mean(), "+/-",scores.std())

Accuracy is:  0.968 +/- 0.02441311123146741


In [49]:
scores

array([0.96, 1.  , 0.95, 0.93, 1.  , 0.94, 0.99, 0.99, 0.97, 0.95])

In [50]:
y_pred = cross_val_predict(model4, X1, np.asarray(y1).reshape(y1.shape[0],), cv=10)
print("Confusion matrix: \n",confusion_matrix(y1,y_pred))

Confusion matrix: 
 [[242   4   8]
 [  4 237  11]
 [  3   2 489]]


In [51]:
y1.groupby("Class")["Class"].count()

Class
EI    254
IE    252
N     494
Name: Class, dtype: int64

In [52]:
print("Error rate for EI: ",(4+8)/254)
print("Error rate for IE: ",(4+11)/252)
print("Error rate for Neither: ",(3+2)/494)

Error rate for EI:  0.047244094488188976
Error rate for IE:  0.05952380952380952
Error rate for Neither:  0.010121457489878543


The experimental results achieved here are more competitive than the results reported on UCI's website.